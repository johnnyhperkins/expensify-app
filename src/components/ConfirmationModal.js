import React from 'react';
import Modal from 'react-modal';

const ConfirmationModal = (props) => (
    <Modal
        isOpen={!!props.isOpen}
        contentLabel="Delete"
        ariaHideApp={false}
        closeTimeoutMS={200}
        onRequestClose={props.handleCloseModal}
        className="modal"
    >
        <h3 className="modal__title">Delete</h3>
        <p className="modal__body">Are you sure you want to delete <strong>{props.expenseDescription}?</strong></p>
        <button className="button" onClick={props.onRemove}>Remove</button>
        <button className="button--secondary button--inline" onClick={props.handleCloseModal}>Cancel</button>
    </Modal>
);

export default ConfirmationModal;