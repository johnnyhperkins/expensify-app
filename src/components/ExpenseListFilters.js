import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        };
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            placeholder="Search expenses"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                            startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                            endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                            endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                            onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                            focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({ 
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()), 
    sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
