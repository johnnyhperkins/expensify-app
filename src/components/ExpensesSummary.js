import React from 'react';
import { connect } from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <p>Displaying {expenseCount} {expenseWord} totalling {numeral( expenseTotal / 100 ).format('$0,0.00')} </p>
        </div>
    );
};

//Whatever object I return in this function is passed as props above. The function below has access to the global state via connect
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: expensesTotal(visibleExpenses)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);