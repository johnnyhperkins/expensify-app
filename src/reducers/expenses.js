const expensesReducerDefaultState = [];

// spread operator returns a copy of an array with whatever new data added to it

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // slightly longer form
            // return state.filter((expense) => expense.id !== action.id );
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id == action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};