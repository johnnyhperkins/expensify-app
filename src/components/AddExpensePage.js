import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        //redirects to the dashboard
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h1>ADD Expense</h1>
                {/* the 'expense' object being passed in here gives us all its properties back (description, amount, note and cratedAt) from the submitted form */}
                <ExpenseForm onSubmit={this.onSubmit} />

                {/* <p>Editing this id: {props.match.params.id}</p> */}
            </div>
        );  
    }
}

const mapDispatchToProps = (dispatch)  => ({ 
    addExpense: (expense) => dispatch(addExpense)
});

//now we have access to props.dispatch
//map dispatch to props does something very similar to mapStateToProps... it adds an object to props that we get to define with the mapDispatchToProps function. It takes in dispatch as an argument and lets us create a method on props that, in this case, takes in the expense object and dispatches it to the store with the addExpense action.

export default connect(undefined, mapDispatchToProps)(AddExpensePage);