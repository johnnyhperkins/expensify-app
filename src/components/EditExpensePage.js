import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense( {id: this.props.expense.id} );
        this.props.history.push('/');  
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit} />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}

//can now access state. to just access dispatch, you only need connect()
const mapStateToProps = (state, props) => {
    return {
        //the compenent above now has access to this exact part of the state, as defined below. We then pass state as an argument to the component above, and that argument is the object which we create (define) here. It could be anything, but in this case it's a single expense. In the component above, the state argument that is passed will be the expense with the matching ID. BUT it also adds it into the props object as props.expense which we define HERE:
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id,expense) => dispatch(editExpense(id,expense)),
        startRemoveExpense: ( data ) => dispatch(startRemoveExpense(data))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);