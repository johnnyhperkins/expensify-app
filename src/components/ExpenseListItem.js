import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = ({description, amount, createdAt, id}) => (
    <div>
        <h3><Link to={"/edit/" + id}>Description: {description}</Link></h3>
        <p>
            {numeral(amount / 100).format('$0,0.00')} 
            - 
            {moment(createdAt).format('MMMM Do YYYY')}
        </p>
    </div>
);

//in this case, connect() gives ExpenseListItem access to the dispatch prop, which is passed into the object the component takes as its argument
export default ExpenseListItem;
