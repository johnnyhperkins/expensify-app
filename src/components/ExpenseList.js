import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
    {
        props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />
        })
        )
    }
        {/* LONG WAY: <ExpenseListItem key={expense.id} description={expense.description} amount={expense.amount} createdAt = {expense.createdAt} /> 
        BETTER WAY: using the spread operator which just takes the expense object and passes it down into ExpenseListItem, which takes an object {description, amount, createdAt} as its argument. But those are only a few of the properties from that object. Also available are the dispatch function and id */}
        
    </div>
);

// you create a function inside of connect() that determines what information from the store you want the component (indicated at the end in parenthesis) to connect or have access to

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//this whole process is simply a way to connect a component to as many or few parts of the store as we want. This line simply says connect the following data from state (store) and pass it to ExpenseLists's props, giving it all the access and benefits
export default connect(mapStateToProps)(ExpenseList);