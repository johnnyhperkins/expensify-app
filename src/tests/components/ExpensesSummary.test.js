import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary component with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={235} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary component with multiple expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={4} expenseTotal={235000} />);
    expect(wrapper).toMatchSnapshot();
});