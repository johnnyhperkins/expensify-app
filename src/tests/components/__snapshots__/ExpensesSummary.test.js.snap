// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary component with 1 expense 1`] = `
<div>
  <p>
    Displaying 
    1
     
    expense
     totalling 
    $2.35
     
  </p>
</div>
`;

exports[`should render ExpensesSummary component with multiple expense 1`] = `
<div>
  <p>
    Displaying 
    4
     
    expenses
     totalling 
    $2,350.00
     
  </p>
</div>
`;
